core/iwasm/common/wasm_c_api.c

```diff
@@ -4711,65 +4633,110 @@ aot_process_export(wasm_store_t *store, const AOTModuleInstance *inst_aot,
 }
 #endif /* WASM_ENABLE_AOT */
 
-wasm_instance_t *
-wasm_instance_new(wasm_store_t *store, const wasm_module_t *module,
-                  const wasm_extern_vec_t *imports, own wasm_trap_t **trap)
-{
-    return wasm_instance_new_with_args(store, module, imports, trap,
-                                       KILOBYTE(32), KILOBYTE(32));
-}
-
 static bool
-compare_imports(const wasm_module_t *module, const wasm_extern_vec_t *imports)
+do_link(const wasm_instance_t *inst, const wasm_module_t *module,
+        const wasm_extern_vec_t *imports)
 {
-    unsigned import_func_count = 0;
-    unsigned import_global_count = 0;
-    unsigned import_memory_count = 0;
-    unsigned import_table_count = 0;
-    unsigned i = 0;
+    uint32 i = 0;
+    uint32 import_func_i = 0;
+    uint32 import_global_i = 0;
 
+    bh_assert(inst && module);
+
+    /*
+     * loader will re-order WASMModule->imports based on item kind
+     * it follows "IMPORT_KIND_FUNC -> IMPORT_KIND_GLOBAL -> IMPORT_KIND_TABLE
+     * -> IMPORT_KIND_MEMORY"
+     *
+     * TODO: it doesn't support to link MEMORY and TABLE for now
+     */
     for (i = 0; imports && i < imports->num_elems; i++) {
         wasm_extern_t *import = imports->data[i];
+
         switch (wasm_extern_kind(import)) {
             case WASM_EXTERN_FUNC:
-                import_func_count++;
+            {
+                link_status ret = LINKED;
+
+                while (ret == LINKED) {
+#if WASM_ENABLE_INTERP != 0
+                    if ((*module)->module_type == Wasm_Module_Bytecode) {
+                        ret = interp_link_func(inst, MODULE_INTERP(module),
+                                               import_func_i,
+                                               wasm_extern_as_func(import));
+                    }
+#endif
+#if WASM_ENABLE_AOT != 0
+                    if ((*module)->module_type == Wasm_Module_AoT) {
+                        ret = aot_link_func(inst, MODULE_AOT(module),
+                                            import_func_i,
+                                            wasm_extern_as_func(import));
+                    }
+#endif
+                    if (ret == UNMATCHED) {
+                        LOG_WARNING("link #%d function failed", import_func_i);
+                        goto failed;
+                    }
+
+                    import_func_i++;
+                }
```

Comment:
Confusing about the link algorithm: why `import_func_i++` on each loop and import_func_i isn't set 0 at the beginning of each loop, it will skip the import func 0 to import_func_i when linking the next import function. And how to ensure their module_name/field_name are the same? Seems here we don't check them. And can the input imports be un-ordered?

My understanding is that: for each imports[i], if it is a function, then we should traversing all the import functions in the wasm module (from module->import_functions[0] to module->import_functions[import_function_count-1]), if module->import_functions[j] hasn't been linked and its module_name/field_name is same with imports[i]'s module_name/field_name, then we can link imports[i] to module->import_functions[j].